src/NUnitTestAdapter/TestConverter.cs

```diff
@@ -127,33 +124,81 @@ public IList<VSTestResult> GetVSTestResults(XmlNode resultNode)
         /// </summary>
         private TestCase MakeTestCaseFromXmlNode(XmlNode testNode)
         {
+            var methodName = testNode.GetAttribute("methodname");
+            var displayName = testNode.GetAttribute("name");
+            var className = testNode.GetAttribute("classname");
+
             var testCase = new TestCase(
                                      testNode.GetAttribute("fullname"),
                                      new Uri(NUnitTestAdapter.ExecutorUri),
                                      _sourceAssembly)
             {
-                DisplayName = testNode.GetAttribute("name"),
+                DisplayName = displayName,
                 CodeFilePath = null,
                 LineNumber = 0
             };
 
             if (_collectSourceInformation && _navigationDataProvider != null)
             {
-                var className = testNode.GetAttribute("classname");
-                var methodName = testNode.GetAttribute("methodname");
                 var navData = _navigationDataProvider.GetNavigationData(className, methodName);
                 if (navData.IsValid)
                 {
                     testCase.CodeFilePath = navData.FilePath;
                     testCase.LineNumber = navData.LineNumber;
+                }               
+            }
+            else
+            {
+                bool isTestNameScenario = IsTestNameScenario(methodName, displayName);
```

Comment:
This line and the method it calls attempts to reverse engineer NUnit in order to decide if the user changed the test name. That's a very fragile thing to do, and it's not clear why you need to do it. Why not __always__ provide the methodname and class name and make use of them by preference to find source information rather than maintaining a hybrid approach in the test explorer?

Of course, for the present, you would have to maintain a hybrid approach anyway, in case the method name and class name are not provided to you, but it makes sense to me that we should drive toward a cleaner solution in the future both for NUnit and other adapters.

Reply:
Generic parameters are not retained in methodname. If we always use classname and methodname then we will not be able to find source information for generic overloaded methods.